build/build.go

```diff
@@ -654,40 +655,25 @@ func (s *Session) BuildPackage(pkg *PackageData) (*compiler.Archive, error) {
 		fmt.Println(pkg.ImportPath)
 	}
 
-	s.Archives[pkg.ImportPath] = archive
-
-	if pkg.PkgObj == "" || pkg.IsCommand() {
-		return archive, nil
-	}
-
-	if err := s.writeLibraryPackage(archive, pkg.PkgObj); err != nil {
-		if strings.HasPrefix(pkg.PkgObj, s.options.GOROOT) {
-			// fall back to first GOPATH workspace
-			firstGopathWorkspace := filepath.SplitList(s.options.GOPATH)[0]
-			if err := s.writeLibraryPackage(archive, filepath.Join(firstGopathWorkspace, pkg.PkgObj[len(s.options.GOROOT):])); err != nil {
-				return nil, err
-			}
-			return archive, nil
-		}
-		return nil, err
-	}
+	s.buildCache.StoreArchive(archive)
+	s.UpToDateArchives[pkg.ImportPath] = archive
 
 	return archive, nil
 }
 
-// writeLibraryPackage writes a compiled package archive to disk at pkgObj path.
-func (s *Session) writeLibraryPackage(archive *compiler.Archive, pkgObj string) error {
-	if err := os.MkdirAll(filepath.Dir(pkgObj), 0777); err != nil {
-		return err
-	}
-
-	objFile, err := os.Create(pkgObj)
-	if err != nil {
-		return err
+// ImportResolverFor returns a function which returns a compiled package archive
+// given an import path. Import paths are resolved relative to the |pkg|.
+func (s *Session) ImportResolverFor(pkg *PackageData) func(string) (*compiler.Archive, error) {
+	return func(path string) (*compiler.Archive, error) {
+		if archive, ok := s.UpToDateArchives[path]; ok {
+			return archive, nil
```

Comment:
I'm probably just not following the logic here, but this appears to check a global cache for `path`, while the GoDoc suggests that `path` should be relative to pkg.

Reply:
Oops, that's a stale comment from an earlier iteration. Fixed.